### PR TITLE
Gpio driver

### PR DESCRIPTION
**Pull request by alexpratte16**
Added GPIO driver files and tests

### CODE REVIEW

File: base/src/gpio/gpio.hpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    void operator = (bool value){
+        mPort->setPinData(mPin, value);
+    }
+
+    bool operator () (){
+        return mPort->getPinData(mPin);
+    }
+
+private:
+    GPIOPort* mPort; //< Ponter to port
```
Comment (carllb) [inline base/src/gpio/gpio.hpp:52]: Spelling mistake

File: base/src/gpio/gpio.hpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+class GPIOController{
+
+public:
+
+    ///
+    /// Get a pointer to the GPIOController instance.
+    /// @return Pointer to the GPIOController instance
+    ///
+    static GPIOController* getInstance(){
+        static int a = 0;
```
Comment (carllb) [inline base/src/gpio/gpio.hpp:66]: what is the variable a for?

File: base/src/gpio/gpio.hpp (hunk 1 above)
Comment (alexpratte16) [reply base/src/gpio/gpio.hpp]: Modified.

File: base/src/gpio/gpio.hpp (hunk 2 above)
Comment (alexpratte16) [reply base/src/gpio/gpio.hpp]: Modified.

File: base/src/gpio/gpio.hpp (hunk 1 above)
Comment (alexpratte16) [reply base/src/gpio/gpio.hpp]: Removed, the variable was for debug purposes.

**Review by carllb — APPROVED** [review]

File: base/src/gpio/gpio.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+/// @param pin pin number (0-15)
+/// @param pull pull to set the pin to.
+///
+void GPIOPort::setPinPull(uint8_t pin, eGPIOPinPull pull){
+    if(pin > 15){
+        return; // invalid pin, do nothing.
+    }
+
+    mPort->PUPDR &= ~(3 << (pin*2)); // clear pull bits.
+    mPort->PUPDR |= (uint8_t)pull << (pin*2);
```
Comment (Christopher-Finelli) [inline base/src/gpio/gpio.cpp:102]: comment this line

**Review by Christopher-Finelli — CHANGES_REQUESTED** [review]
one missing comment line 102 of gpio.cpp

File: base/src/gpio/gpio.cpp (hunk 3 above)
Comment (alexpratte16) [reply base/src/gpio/gpio.cpp:102]: Added comment

**Review by Christopher-Finelli — APPROVED** [review]

**Pull request merged**